File: leetcode/blind_75_must_do/word_search.go
/*
Given an m x n grid of characters board and a string word, return true if word exists in the grid.

The word can be constructed from letters of sequentially adjacent cells, where adjacent cells are
horizontally or vertically neighboring. The same letter cell may not be used more than once.

Example 1:
Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "ABCCED"
Output: true

Example 2:
Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "SEE"
Output: true

Example 3:
Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "ABCB"
Output: false
*/

package main

import "fmt"

var (
	row, col int
	visited  = [][]bool{}
)

func dfs(matrix [][]byte, word string, idx, i, j int) bool {
	if idx == len(word) {
		return true
	}

	if i < 0 || j < 0 || i >= row || j >= col {
		return false
	}

	if visited[i][j] {
		return false
	}

	if matrix[i][j] != word[idx] {
		return false
	}

	visited[i][j] = true
	res := dfs(matrix, word, idx+1, i+1, j) ||
		dfs(matrix, word, idx+1, i-1, j) ||
		dfs(matrix, word, idx+1, i, j+1) ||
		dfs(matrix, word, idx+1, i, j-1)
	visited[i][j] = false

	return res

}

func word_search(matrix [][]byte, word string) bool {
	if len(word) == 0 {
		return false
	}

	row, col = len(matrix), len(matrix[0])

	visited = make([][]bool, row)
	for i := 0; i < row; i++ {
		visited[i] = make([]bool, col)
	}

	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			if dfs(matrix, word, 0, i, j) {
				return true
			}
		}
	}
	return false
}

func main() {
	fmt.Println(word_search([][]byte{
		{'A', 'B', 'C', 'E'},
		{'S', 'F', 'C', 'S'},
		{'A', 'D', 'E', 'E'},
	}, "ABCCED"))

}

File: LeetCode/Medium/Word Search/Cpp/word_search.cc
#include <iostream>
#include <vector>

typedef std::vector<char> vc;
typedef std::vector<vc> vcc;
typedef std::vector<bool> vb;
typedef std::vector<vb> vbb;

bool dfs(std::string word, int i, int j, int k, vbb &visited, vcc &board) {
    if (k == word.length()) return true;
    if (i<0 || j<0 || i>=board.size() || j>=board[0].size() || visited[i][j]) return false;
    if (word[k] != board[i][j]) return false;
    visited[i][j] = true;
    bool res = 
        dfs(word, i+1, j, k+1, visited, board) ||
        dfs(word, i-1, j, k+1, visited, board) ||
        dfs(word, i, j+1, k+1, visited, board) ||
        dfs(word, i, j-1, k+1, visited, board);
    visited[i][j] = false;
    return res;
}

bool word_search(vcc &board, std::string word) {
    int n = board.size(), m = board[0].size();
    vbb visited(n, vb(m));
    for (int i=0;i<n;i++) {
        for (int j=0;j<m; j++) {
            if (board[i][j] == word[0]) {
                if (dfs(word, i, j, 0, visited, board)) {
                    return true;
                }
            }
        }
    }
    return false;
}

int main () {
    vcc v{{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}};
    std::cout<<word_search(v, "ABCCED")<<"\n";
}
